### PR TITLE
Add prepare inputs utility for AutoDock Vina

### PR DESCRIPTION
**Pull request by ncfrey**
This PR adds a short utility function for preparing protein-ligand complex inputs to AutoDock Vina. Vina requires sanitized inputs saved as pdb files. Currently, this utility does the following:
* Reads PDB file or a PDBID for a protein and downloads from PDB
* Reads PDB file or a SMILES string for a ligand 
* Uses `PDBFixer` to fix issues in the protein PDB file
* Uses RDKit to optimize the ligand geometry 
* Sanitizes both molecules with RDKit and returns as RDKit Molecules 
* Optionally writes prepared molecules to PDB files for Vina

I know there is a lot of complexity involved in preparing complexes, this is just what worked for me to get `dc.dock` to work. I'm hoping this will also standardize how inputs are prepared for docking and ML workflows in DeepChem. Happy to make any suggested changes.

CCing @peastman and @rbharath

### CODE REVIEW

**Comment by peastman** [comment]
What you're doing here mostly looks good.  But it should never be treated as a black box.  Fixing up PDB files can be complicated.  PDBFixer tries to automate as much as it can and make reasonable best guesses, but a human should always inspect the results.  Maybe the PDB file is a dimer, and one of the two copies should be removed before you try to dock anything to it.  Maybe the file includes molecules that were added to help the protein crystallize, and they should be removed.  Maybe it has large loops that weren't resolved in the crystal structure.  PDBFixer will fill in *something* for them, but not necessarily anything very realistic.  Maybe there are long tails at the ends of the chain that it would be better to just omit.  Human judgement is required to produce a structure suitable for docking.

File: deepchem/utils/vina_utils.py [hunk 1]
```diff
@@ -114,3 +114,72 @@ def load_docked_ligands(
...
+  pdb_name: Optional[str]
+    If given, write sanitized protein and ligand to files called
+    "pdb_name.pdb" and "ligand_pdb_name.pdb"
+
+  Returns
+  -------
+  Tuple[RDKitMol, RDKitMol]
+    Tuple of `protein_molecule, ligand_molecule` with 3D information.
+
+  Notes
```
Comment (rbharath) [inline deepchem/utils/vina_utils.py:142]: s/Notes/Note since that displays better on the docs :)

File: deepchem/utils/vina_utils.py [hunk 2]
```diff
@@ -114,3 +114,72 @@ def load_docked_ligands(
...
+  try:
+    from rdkit import Chem
+    from pdbfixer import PDBFixer
+    from simtk.openmm.app import PDBFile
+  except ModuleNotFoundError:
+    raise ImportError(
+        "This function requires RDKit and OpenMM to be installed.")
+
+  if protein.endswith('.pdb'):
+    fixer = PDBFixer(protein)
```
Comment (rbharath) [inline deepchem/utils/vina_utils.py:158]: Oh this is slick! Didn't know `PDBFixer` could pull down from a URL :)

File: deepchem/utils/vina_utils.py [hunk 3]
```diff
@@ -114,3 +114,72 @@ def load_docked_ligands(
...
+  else:
+    fixer = PDBFixer(url='https://files.rcsb.org/download/%s.pdb' % (protein))
+
+  if ligand.endswith('.pdb'):
+    m = Chem.MolFromPDBFile(ligand)
+  else:
+    m = Chem.MolFromSmiles(ligand, sanitize=True)
+
+  # Apply common fixes to PDB files
+  fixer.findMissingResidues()
```
Comment (rbharath) [inline deepchem/utils/vina_utils.py:168]: One option to make this a little more white-box as @peastman suggests might be to perhaps add keyword arguments that allow users to control which of these transformations are applied

File: deepchem/utils/vina_utils.py [hunk 4]
```diff
@@ -114,3 +114,72 @@ def load_docked_ligands(
...
+  fixer.findNonstandardResidues()
+  fixer.replaceNonstandardResidues()
+  fixer.removeHeterogens(False)  # remove water
+  fixer.addMissingHydrogens(7.0)
```
Comment (rbharath) [inline deepchem/utils/vina_utils.py:172]: The pH should probably be another keyword argument, same with whether to remove water as a heterogen

File: deepchem/utils/vina_utils.py [hunk 5]
```diff
@@ -114,3 +114,72 @@ def load_docked_ligands(
...
+  PDBFile.writeFile(fixer.topology, fixer.positions, open('tmp.pdb', 'w'))
+  p = Chem.MolFromPDBFile('tmp.pdb', sanitize=True)
+
+  # Optimize ligand
+  m = Chem.AddHs(m)  # need hydrogens for optimization
```
Comment (rbharath) [inline deepchem/utils/vina_utils.py:177]: Probably a couple more keyword arguments here :)

File: deepchem/utils/vina_utils.py [hunk 6]
```diff
@@ -114,3 +114,72 @@ def load_docked_ligands(
...
+
+
+def prepare_inputs(protein: str, ligand: str,
+                   pdb_name: Optional[str] = None) -> Tuple[RDKitMol, RDKitMol]:
+  """This prepares protein-ligand complexes for docking.
+
+  Autodock Vina requires PDB files for proteins and ligands with
+  sensible inputs. This function uses PDBFixer and RDKit to ensure
+  that inputs are reasonable and ready for docking.
+
```
Comment (rbharath) [inline deepchem/utils/vina_utils.py:126]: Could you add an examples section here with some usage examples showing how to process a few different protein/ligand systems using this function?

**Review by rbharath — COMMENTED** [review]
I like this PR! It'll be very useful as we make our docking/structural infrastructure more powerful. I have a few suggestions below for how to make this a little less black box as @peastman suggests above

**Comment by ncfrey** [comment]
> What you're doing here mostly looks good. But it should never be treated as a black box. Fixing up PDB files can be complicated. PDBFixer tries to automate as much as it can and make reasonable best guesses, but a human should always inspect the results.

Ok great, thanks for the comments! I've added arguments and additional notes to make it clear that the results should be inspected.

File: deepchem/utils/vina_utils.py (hunk 2 above)
Comment (ncfrey) [reply deepchem/utils/vina_utils.py:192]: Yeah it's very handy!

File: deepchem/utils/vina_utils.py (hunk 3 above)
Comment (ncfrey) [reply deepchem/utils/vina_utils.py]: Done, added a bunch of arguments to set defaults and control the transformations.

File: deepchem/utils/vina_utils.py (hunk 6 above)
Comment (peastman) [reply deepchem/utils/vina_utils.py:136]: How about this:

> Autodock Vina requires PDB files for proteins and ligands with sensible inputs. This function uses PDBFixer and RDKit to attempt to produce inputs that are reasonable and ready for docking. Default values are given for convenience, but fixing PDB files is complicated and human judgement is required to produce protein structures suitable for docking.  Always inspect the results carefully before trying to perform docking.

Comment (ncfrey) [reply deepchem/utils/vina_utils.py:136]: Sounds great, I'll add that in!

**Review by rbharath — APPROVED** [review]
This looks quite solid! I think this is ready to merge in on my end. Will mark this as approved, but will hold off on merging in case @peastman has more comments

**Comment by peastman** [comment]
Aside from the text change I suggested it looks fine.

**Comment by ncfrey** [comment]
Added @peastman's suggested text change!

**Comment by rbharath** [comment]
Ok sounds good! Feel free to merge in once ready :)

**Comment by peastman** [comment]
This PR introduced a flake8 error that is now causing travis to fail on the main branch.  See #2297.  @ncfrey can you fix it?

**Pull request merged**